Shopify/cli
# Adopt `pnpm` as the default package manager

### WHY are these changes introduced?
Despite Yarn having done a reasonably good job at managing the dependencies of the project, there are a few issues rooted in their design that might manifest on the user side and worse the experience contributing to the repo. In particular:

- Yarn's **hoisting** mechanism combined with how the Node module system works might lead to packages depending on `dependencies` that are not declared as their dependencies. This can go unnoticed when building and testing the packages locally and on CI. Some projects prefer to use [ESLint](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-extraneous-dependencies.md) to catch these issues.
- Yarn's management of the **global cache** and the IO operations (copying and pasting files from the cache into `node_modules`) proved to be inconsistently unreliable leading to `yarn install` failures that cause developers to either retry a build on CI or deleting `node_modules` locally and trying again.


### WHAT is this pull request doing?

I'm replacing Yarn with PNPM.

[PNPM](https://pnpm.io/) addresses the above issues first, by [being more strict](https://www.kochan.io/nodejs/pnpms-strictness-helps-to-avoid-silly-bugs.html) with the directory structure and the usage of `peerDependencies`, and second, by having a global cached that's accessed through [symlinks](https://pnpm.io/symlinked-node-modules-structure). PNPM's has a more sophisticated caching system with a content-addressable store that leads to faster dependency installation.

Even though the above issues might seem minor, they'll become more frequent as our monorepo grows. I'm confident this is a great step forward for contributors and users. Other projects like Vite or Vitest have already validated the tool and use it at a much larger scale than us.

### Why not Yarn 2/3?
They made [plug & play](https://yarnpkg.com/features/pnp) the default approach to expose the module graph to the runtime and it's [not compatible with ESM](https://github.com/yarnpkg/berry/issues/1149) yet. The API they'd need to build upon, [loaders](https://nodejs.org/api/esm.html#loaders), it's flag as experimental in Node. Moreover that approach is not conventional and therefore requires additional setup of our transpilation tools and workflows to make them work. This will make the project harder to reason about.

### How to test your changes?
1. Check out the branch
2. Run  `rm -rf node_modules/ **/node_modules` & `pnpm install`
3. Run `pnpm build && pnpm lint && pnpm test`
4. `yarn` should output a message telling you to use `pnpm` instead.


### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
